### PR TITLE
[FLINK-25277] add shutdown hook to stop TaskExecutor on SIGTERM

### PR DESCRIPTION
**Pull request by metaswirl**
## What is the purpose of the change

Jira Issue: https://issues.apache.org/jira/browse/FLINK-25277

In short:
- We add graceful termination for the TaskExecutor on  SIGTERM
- We synchronize the termination of the YARNResourceManagerDriver with the concurrent release of containers

## Brief change log

- Add shutdown hook to TaskManagerRunner
- Add java.util.concurrent.Phaser to YARNResourceManagerDriver
- Synchronize on TaskExecutorStateChangelogStoragesManager as it is accessed by multiple shutdown hooks (it's own and the one from TaskExecutor)

## Verifying this change

The YARN tests cover the setup and shutdown of clusters. Particular org.apache.flink.yarn.YARNSessionFIFOITCase.

After fixing the synchronization issue (discussed in far greater detail on Jira), we ran the test for 100 iterations on the Azure Pipelines.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: yes
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

**Comment by metaswirl** [comment]
@flinkbot run azure

**Comment by metaswirl** [comment]
@flinkbot run azure

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManagerDriver.java [hunk 1]
```diff
@@ -684,6 +694,7 @@ public void onGetContainerStatusError(ContainerId containerId, Throwable throwab
 
         @Override
         public void onStopContainerError(ContainerId containerId, Throwable throwable) {
+            trackerOfReleasedResources.arriveAndDeregister();
```
Comment (tillrohrmann) [inline flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManagerDriver.java:697]: What if a container errors that is not released? Wouldn't this thwart our bookkeeping?

**Review by tillrohrmann — COMMENTED** [review]
Thanks for updating this PR @metaswirl. I had a question concerning the correctness of the bookkeeping.

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManagerDriver.java (hunk 1 above)
Comment (metaswirl) [reply flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManagerDriver.java:697]: Not sure that I follow. If we start the container release process we basically count up (register). When the container release process completes successfully we count down (deregister). The same happens, when the release process fails (onStopContainerError). If we start a new release process, we would still count up at the start and count down at the end.

Are you saying that scenarios exist where we count up, but never count down or vice versa? This can only happen, if the callback is for some reason never called. Possibly due to a failure in the YARN clients. Even in this (unlikely?) case, the containers will be killed shortly afterwards anyway. (That is, if the shutdown procedure is initated over YARN's kill application command.)

**Review by tillrohrmann — APPROVED** [review]
Thanks for explaining how the Yarn logic works @metaswirl. I think this is a good solution and works. One last question is whether we want to add an IT test that ensures that a TM deregisters from a `JM`.

**Review by dmvk — CHANGES_REQUESTED** [review]
Thanks for the PR @metaswirl, it looks good to me overall! 🎉 My main concern about the shutdown hook lifecycle, PTAL.

Also I would probably choose the Phaser approach as I find the interface kind of confusing, by no objections there.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskExecutorStateChangelogStoragesManager.java [hunk 2]
```diff
@@ -124,10 +124,12 @@ public void releaseStateChangelogStorageForJob(@Nonnull JobID jobId) {
     }
 
     public void shutdown() {
-        if (closed) {
-            return;
+        synchronized (this) {
```
[the remark this answers is not in the document]
Comment (metaswirl) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskExecutorStateChangelogStoragesManager.java:128]: good idea!

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManagerDriver.java [hunk 3]
```diff
@@ -117,6 +118,8 @@
     private TaskExecutorProcessSpecContainerResourcePriorityAdapter
             taskExecutorProcessSpecContainerResourcePriorityAdapter;
 
+    private Phaser trackerOfReleasedResources;
```
[the remark this answers is not in the document]
Comment (metaswirl) [reply flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManagerDriver.java]: thanks :)

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java [hunk 4]
```diff
@@ -106,6 +107,8 @@
     private static final int SUCCESS_EXIT_CODE = 0;
     @VisibleForTesting static final int FAILURE_EXIT_CODE = 1;
 
+    private static Thread shutdownHook;
```
[the remark this answers is not in the document]
Comment (metaswirl) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java]: Good points all around, thank you! I've changed the implementation accordingly.

**Comment by metaswirl** [comment]
> Also I would probably choose the Phaser approach as I find the interface kind of confusing, by no objections there.

Do you have an alternative approach in mind? (locks and counter?) I am not particularly attached to the Phaser approach. Its advantage is that it can be used for this problem without requiring additional objects. Its disadvantage is that it covers a far greater problem space and so the method names are a bit misleading for this scenario. (register -> count up, arriveAndDeregister -> count down, arriveAndAwaitAdvance -> countDown and wait for count to reach 0 before continuing)

**Comment by metaswirl** [comment]
> Thanks for explaining how the Yarn logic works @metaswirl. I think this is a good solution and works. One last question is whether we want to add an IT test that ensures that a TM deregisters from a `JM`.

I tried testing this. But, in my scenarios, the TaskManager often received the SIGTERM after the ResourceManager had already exited. I tried to artifcially delay the termination of the ResourceManager. But I must have done it in a wrong way, because the message from the TaskManager was not received by the RM. (I saw the Akka message being sent, but not received.)

I added tests that check the termination completes as long as (a) no resources are simultaneously released, (b) resources are successfully simultaneously released and (c) resources are unsuccessfully simultaneously released.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskExecutorStateChangelogStoragesManager.java [hunk 5]
```diff
@@ -124,10 +125,11 @@ public void releaseStateChangelogStorageForJob(@Nonnull JobID jobId) {
...
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskExecutorStateChangelogStoragesManager.java:128]: I think we don't need `synchronized` if we are using `AtomicBoolean` here.

**Review by tillrohrmann — COMMENTED** [review]
Thanks for updating this PR @metaswirl. Concerning the test I was more thinking about a test where we kill a TM process via `SIGTERM` and check that it shuts down (e.g. deregisters from a `JM`).

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskExecutorStateChangelogStoragesManager.java (hunk 5 above)
Comment (metaswirl) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskExecutorStateChangelogStoragesManager.java]: yes, of course.

**Comment by metaswirl** [comment]
> Thanks for updating this PR @metaswirl. Concerning the test I was more thinking about a test where we kill a TM process via `SIGTERM` and check that it shuts down (e.g. deregisters from a `JM`).

So this is not so much about the whole YARN thing, but about the shutdown hook itself, correct? To test this, we would need to (a) catch the message to the JobManager, (b) replace the JobManager by a mock or (c) access the JobManager's state via the REST interface. This should be doable. I will have to look, if there are already tests for this.

**Comment by tillrohrmann** [comment]
Quick comment because I see a merge commit: The Flink community uses rebasing to have a linear history.

**Comment by dmvk** [comment]
> Do you have an alternative approach in mind? (locks and counter?) I am not particularly attached to the Phaser approach. Its advantage is that it can be used for this problem without requiring additional objects. Its disadvantage is that it covers a far greater problem space and so the method names are a bit misleading for this scenario. (register -> count up, arriveAndDeregister -> count down, arriveAndAwaitAdvance -> countDown and wait for count to reach 0 before continuing)

I'd probably go with some kind of tracking mechanism based on the `ConcurrentMap` + `CompletableFutures`, which of course have a higher memory footprint, but should be easier to reason about / debug (this might be bit biased).

The current approach is good, we should keep it.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskExecutorStateChangelogStoragesManager.java [hunk 6]
```diff
@@ -124,10 +125,9 @@ public void releaseStateChangelogStorageForJob(@Nonnull JobID jobId) {
     }
 
     public void shutdown() {
-        if (closed) {
+        if (closed.getAndSet(true)) {
             return;
         }
-        closed = true;
 
         HashMap<JobID, Optional<StateChangelogStorage<?>>> toRelease =
```
[the remark this answers is not in the document]
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskExecutorStateChangelogStoragesManager.java:132]: Thanks @dmvk, I've created FLINK-25678 to fix this.

**Comment by tillrohrmann** [comment]
@dmvk can this change be merged? I would need it for a fix.

**Comment by dmvk** [comment]
@tillrohrmann There is one small thing missing that we've discussed offline and that is replacing the busy waiting loop in the IT case.

**Comment by tillrohrmann** [comment]
Alright, no hurries. I realized that I needed more changes than what this PR introduces to the `TaskManagerRunner`. We probably will have some merge conflicts in this area once #18419 gets merged.

**Comment by metaswirl** [comment]
> You could take a look at `AbstractTaskManagerProcessFailureRecoveryTest`. I think what could work is to start a `StandaloneSessionClusterEntrypoint` that starts a `RM`. Then you could set the heartbeat timeout and number of missed heartbeat messages to something high so that you don't see heartbeat timeouts. Last but not least you could check whether the TM is no longer registered at the RM by querying the set of registered TMs.

Thanks, we did something very similar to what you proposed.

File: flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerDisconnectOnShutdownITCase.java [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+        String javaCommand = getJavaCommandPath();
+        if (javaCommand == null) {
+            fail("cannot find java executable");
+        }
+
+        TaskManagerConnectionTracker tracker = new TaskManagerConnectionTracker();
+
+        TestProcessBuilder.TestProcess taskManagerProcess = null;
+        try (final StandaloneSessionClusterEntrypoint clusterEntrypoint =
+                new StandaloneSessionClusterEntrypoint(config) {
```
Comment (dmvk) [inline flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerDisconnectOnShutdownITCase.java:105]:
```suggestion
        try (final SessionClusterEntrypoint clusterEntrypoint =
                new SessionClusterEntrypoint(config) {
```

File: flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerDisconnectOnShutdownITCase.java [hunk 8]
```diff
@@ -0,0 +1,304 @@
...
+        config.set(JobManagerOptions.RESOURCE_WAIT_TIMEOUT, Duration.ofSeconds(30L));
+
+        // check that we run this test only if the java command
+        // is available on this machine
+        String javaCommand = getJavaCommandPath();
+        if (javaCommand == null) {
+            fail("cannot find java executable");
+        }
+
+        TaskManagerConnectionTracker tracker = new TaskManagerConnectionTracker();
```
Comment (dmvk) [inline flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerDisconnectOnShutdownITCase.java:101]: nit: just a good practice to follow, always mark all the variables as `final` unless you intend to mutate them. It's bit more verbose, but it gives a good hint to anyone reading the code and it's less error prone in general.

File: flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerDisconnectOnShutdownITCase.java [hunk 9]
```diff
@@ -0,0 +1,304 @@
...
+        }
+    }
+
+    // --------------------------------------------------------------------------------------------
+
+    /**
+     * The entry point for the TaskExecutor JVM. Simply configures and runs a TaskExecutor. Copy
+     * from AbstractTaskManagerProcessFailureRecoveryTest.
+     */
+    public static class TaskExecutorProcessEntryPoint {
```
Comment (dmvk) [inline flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerDisconnectOnShutdownITCase.java:178]: again, can we reuse the one from `AbstractTaskManagerProcessFailureRecoveryTest`?

File: flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerDisconnectOnShutdownITCase.java [hunk 10]
```diff
@@ -0,0 +1,304 @@
...
+import org.apache.flink.runtime.rpc.FatalErrorHandler;
+import org.apache.flink.runtime.rpc.RpcService;
+import org.apache.flink.runtime.taskexecutor.SlotReport;
+import org.apache.flink.runtime.taskexecutor.TaskExecutorResourceUtils;
+import org.apache.flink.runtime.taskexecutor.TaskManagerRunner;
+import org.apache.flink.test.util.TestProcessBuilder;
+import org.apache.flink.util.ConfigurationException;
+import org.apache.flink.util.TestLogger;
+
+import org.junit.Test;
```
Comment (dmvk) [inline flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerDisconnectOnShutdownITCase.java:57]: we should use junit5 for new unit tests

File: flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerDisconnectOnShutdownITCase.java [hunk 11]
```diff
@@ -0,0 +1,304 @@
...
+import java.util.concurrent.Executor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.apache.flink.runtime.testutils.CommonTestUtils.getJavaCommandPath;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.fail;
+
+/** This test ensures the TaskManager disconnects from the ResourceManager on shutdown. */
+public class TaskManagerDisconnectOnShutdownITCase extends TestLogger {
```
Comment (dmvk) [inline flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerDisconnectOnShutdownITCase.java:75]: with junit5 we have an junit extension for the testlogger

File: flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerDisconnectOnShutdownITCase.java [hunk 12]
```diff
@@ -0,0 +1,304 @@
...
+
+import java.time.Duration;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.apache.flink.runtime.testutils.CommonTestUtils.getJavaCommandPath;
+import static org.junit.Assert.assertEquals;
```
Comment (dmvk) [inline flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerDisconnectOnShutdownITCase.java:71]: also the community has agreed upon unifying the assertion framework, so we should stick with using assertj for the new tests

File: flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerDisconnectOnShutdownITCase.java [hunk 13]
```diff
@@ -0,0 +1,304 @@
...
+                if (!taskManagerProcess.getProcess().waitFor(30, TimeUnit.SECONDS)) {
+                    log.error("TaskManager did not shutdown in time.");
+                    printProcessLog(taskManagerProcess);
+                    taskManagerProcess.getProcess().destroyForcibly();
+                }
+            }
+        }
+    }
+
+    protected static void printProcessLog(TestProcessBuilder.TestProcess process) {
```
[the remark this answers is not in the document]
Comment (metaswirl) [reply flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerDisconnectOnShutdownITCase.java:156]: This is a slightly modified version that doesn't require a name argument. We can of course re-use the method, but it is slightly more verbose and the cross dependency doesn't make this test easier to understand. I'd rather keep it as is.

File: flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerDisconnectOnShutdownITCase.java [hunk 14]
```diff
@@ -0,0 +1,304 @@
...
+                    tracker.getNumberOfConnectedTaskManager());
+        } catch (Exception e) {
+            e.printStackTrace();
+            printProcessLog(taskManagerProcess);
+            fail(e.getMessage());
+        } catch (Error e) {
+            e.printStackTrace();
+            printProcessLog(taskManagerProcess);
+            fail(e.getMessage());
+            throw e;
```
[the remark this answers is not in the document]
Comment (metaswirl) [reply flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerDisconnectOnShutdownITCase.java:143]: Sounds good. I see you also removed the `fail` method call. Was this on purpose?

File: flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerDisconnectOnShutdownITCase.java [hunk 15]
```diff
@@ -0,0 +1,304 @@
...
+        } finally {
+            if (taskManagerProcess != null) {
+                taskManagerProcess.destroy();
```
[the remark this answers is not in the document]
Comment (metaswirl) [reply flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerDisconnectOnShutdownITCase.java:146]: yes, of course. will change.

File: flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerDisconnectOnShutdownITCase.java [hunk 16]
```diff
@@ -0,0 +1,304 @@
...
+        private final AtomicInteger numberOfConnectedTaskManager = new AtomicInteger();
+
+        public void connectTaskRanager() {
+            if (!taskManagerConnectedFuture.isDone()) {
+                taskManagerConnectedFuture.complete(null);
+            }
+            numberOfConnectedTaskManager.incrementAndGet();
+        }
+
+        public void disconnectTaskRanager() {
```
[the remark this answers is not in the document]
Comment (metaswirl) [reply flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerDisconnectOnShutdownITCase.java:286]: thx!

File: flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerDisconnectOnShutdownITCase.java (hunk 7 above)
Comment (metaswirl) [reply flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerDisconnectOnShutdownITCase.java:105]: nice, I didn't consider this.

File: flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerDisconnectOnShutdownITCase.java (hunk 8 above)
Comment (metaswirl) [reply flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerDisconnectOnShutdownITCase.java:101]: true

File: flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerDisconnectOnShutdownITCase.java (hunk 9 above)
Comment (metaswirl) [reply flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerDisconnectOnShutdownITCase.java:178]: how about we refactor AbstractTaskManagerProcessFailureRecoveryTest into a standalone class?

File: flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerDisconnectOnShutdownITCase.java (hunk 11 above)
Comment (metaswirl) [reply flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerDisconnectOnShutdownITCase.java]: done

File: flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerDisconnectOnShutdownITCase.java (hunk 12 above)
Comment (metaswirl) [reply flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerDisconnectOnShutdownITCase.java]: done

**Review by dmvk — APPROVED** [review]
LGTM 👍 Thanks for the PR @metaswirl, good job!

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java [hunk 17]
```diff
@@ -179,6 +179,10 @@ public TaskManagerRunner(
                         () -> this.closeAsync(Result.JVM_SHUTDOWN).join(),
                         getClass().getSimpleName(),
                         LOG);
+        MemoryLogger.startIfConfigured(
```
Comment (dmvk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java:182]: is this related?

**Comment by metaswirl** [comment]
This change no longer adds the shutdown hook, as this has already been done by commit: dd6069fabf8a7ff65fbd9ff8dd7b0c47f492288f.

**Review by dmvk — APPROVED** [review]
👍

**Comment by tillrohrmann** [comment]
Sorry for the inconveniences I've caused with my merge. It caused just more trouble. 🙏

**Comment by metaswirl** [comment]
> Sorry for the inconveniences I've caused with my merge. It caused just more trouble. 🙏

Thank you for fixing the code 👍
